Fix potential issues with medal display

peppy opened the pull request:
This is mostly a refactor of the component to bring code quality up to 2024 standards. It's still a tad finicky.

Checking the logs, the online texture load never occurred after the preparation call (there's no log of the web request even starting). This suggests the overlay was somehow hidden before the `LoadComponentAsync` could be acted upon. Maybe the overlay was hidden by external forces?

Fixes probably https://github.com/ppy/osu/issues/30626 but I haven't been able to repro the underlying failure.

I also took the opportunity to decouple the loading of the medal sprite from the animation. While this was likely not the underlying failure, I'd rather a resource stuck in a loading state didn't leave the game in a weird state. Previously, it could have caused the precise issue in the linked issue (can be tested by adding a sleep in the `load()` method).

peppy committed:
Ensure `DrawableMedal` loading doesn't ever block on online resources

peppy committed:
Refactor `MedalOverlay` to be more readable

Shouldn't really have any functionality changes, just fixing some old
code that I can't easily parse these days.

peppy committed:
Better control of show/hide of overlay

peppy committed:
Fix hidden dissmissing logic

smoogipoo commented on osu.Game/Overlays/MedalSplash/DrawableMedal.cs:
```diff
@@ -122,11 +121,12 @@ public DrawableMedal(Medal medal)
             medalGlow.Texture = textures.Get(@"MedalSplash/medal-glow");
             description.Colour = colours.BlueLight;
+
+            Logger.Log("loaded");
```
Pretty non-descript log message.

peppy replied:
oops, was debug

smoogipoo commented on osu.Game/Overlays/MedalSplash/DrawableMedal.cs:
```diff
@@ -51,7 +50,7 @@ public DrawableMedal(Medal medal)
                     Children = new Drawable[]
                     {
-                        medalSprite = new Sprite
+                        new DelayedLoadWrapper(() => new MedalOnlineSprite(medal), 0)
```
Is this really what we want? Doesn't this mean the medal will display without a texture on slow internet? I would rather it were delayed (by LCA) as it was previously. The LCA and whole "preparing to display" stuff in `MedalOverlay` looks unnecessary now because it's not doing anything that would take a long time to load?

peppy replied:
> The LCA and whole "preparing to display" stuff in MedalOverlay looks unnecessary now because it's not doing anything that would take a long time to load?

It's still loading a sample and texture. If a stutter from that is not of concern, then I guess it could be removed, but I'd prefer having it there?

smoogipoo replied:
I don't understand why this is better than having the whole overlay sit there loading? I would rather have the overlay _look good_ than to have it display immediately as soon as a medal is awarded, no?

Master:

https://github.com/user-attachments/assets/d17a1d13-1546-4dcb-9be4-246cfbba5a77

This PR:

https://github.com/user-attachments/assets/e6fc3d62-442b-41ed-b715-14e9681be9dd

> If a stutter from that is not of concern

Have you actually noticed a stutter there...? I'd be highly concerned if that caused a stutter because we do that pattern everywhere else in the game without LCA.

smoogipoo replied:
Dunno. Perhaps I'm being dense with the above, so will ask for a second opinion from @bdach

peppy replied:
I mean, anywhere we are loading sprites or samples can cause a stutter, right? So in general we try and load these asynchronously?

I can revisit this, but it'll probably be a further refactor as the overlay simply cannot be present while the load is happening (right now on `master` it blocks the whole game for the *entire duration of the resource load*).

bdach replied:
> right now on `master` it blocks the whole game for the _entire duration of the resource load_

I might be talking out of rear here, but isn't that possibly fixable by delaying the show here?:

https://github.com/ppy/osu/blob/1e6c04e98b092e52a35b20d07e9a5a67e61de1b4/osu.Game/Overlays/MedalOverlay.cs#L155-L156

As in, move the `Show()` call to the LCA callback there.

My general opinion is that it'd obviously be better to not use DLW as it'll inevitably incur a higher risk ot things looking jank than ensuring everything is loaded prior to show, but I wouldn't necessarily insist on it if it turns out to be too annoying to make happen.

peppy replied:
Yeah, potentially that works. Just have to make sure the load actually still works without the thing loading being present. It would be the refactor I'd look at doing.

I'll give that direction a try and see how it goes. It kind of freaks me out that right now, a loading texture can block medals from ever displaying again (I'm not sure if we have an established "load failed" path for drawables?).

smoogipoo replied:
I don't think that will work because the scheduler (LCA callback) is only updated if the overlay is present. But there's like 1001 different ways to get around that, not the least being to use the scheduler callback on LCA, set AlwaysPresent, or override IsPresent.

smoogipoo replied:
To be honest if that's so hard to handle and DLW is used as a roundabout way to handle that, then I'm even less confident about this change. I'm not interested in quick merging this as it was reported once ever, so I would rather it be fixed properly as long as we agree that my dislike of DLW is not insane.

Also I'm not exactly sure what can "fail" here, `OnlineStore` try-catches everything and returns `null`, and so will `TextureStore`. I understand you haven't reproed it (and I haven't either - tested prior to this PR), but that feels like a weird thing to hinge on.

peppy committed:
Remove debug logging

peppy committed:
Revert "Ensure `DrawableMedal` loading doesn't ever block on online resources"

This reverts commit 8585327858a00b6c15612bf29e446ccb733773d9.

peppy committed:
Change point of queueing to avoid loading-from-in-queue

peppy committed:
Simplify hide code by moving to common method

bdach committed:
Attempt to fix test failures

bdach committed:
Fix LCA call crashing in actual usage

It's not allowed to call `LoadComponentsAsync()` on a background thread:

	https://github.com/ppy/osu-framework/blob/fd64f2f0d47f0ee1aaa596bde1e83e527d610340/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L147

and in this case the event that causes the LCA call is dispatched from a
websocket client, which is not on the update thread, so scheduling is
required.

bdach reviewed (approved):
I tried to fix test failures with https://github.com/ppy/osu/pull/30877/commits/b0958c8d418db28022fe2d12dd0ca2722ddad14c (seemingly successfully) and also determined that this would die on an `InvalidThreadForMutationException` in full stack testing with a web instance, which I also fixed in https://github.com/ppy/osu/pull/30877/commits/c14fe21219acc24741b7d6b31106763fdd488796.

Looks good to me otherwise, but I'll request a second check ---

bdach commented:
@peppy probably go over my changes as well whenever able.

peppy commented:
Hmm, I considered thread safety of LCA but was 100% convinced it's fine from a background thread. TIL.

Changes look good, thanks.

bdach commented:
This has been laying around approved for a while now so I'm just gonna merge and see what happens.

bdach committed:
Merge branch 'master' into medal-display-fix
